### PR TITLE
First stab at improving FK constraint handling in data insertion and del...

### PR DESCRIPTION
**Pull request by neilbmclaughlin**
First stab at improving FK constraint handling in data insertion and deletion. The StubDatabaseutility now keeps an ordered record of insertions and for delete just runs backwards through the insert list

### CODE REVIEW

File: NHSChoices.StubDatabaseUtility.Tests.Integration/StubDatabaseTests.cs [hunk 1]
```diff
@@ -37,47 +65,136 @@ public void DatabaseCopy()
...
     {
-      return new StubDatabaseUtility(serverConnection);
+      //Arrange
+
+      //Act + Assert
+      var exception = Assert.Throws<Exception>(() => _stubDatabaseUtility.DeleteTableData());
+      Assert.That(exception.Message, Is.EqualTo("Delete from table 'UnknownTable' failed. See inner exception for details."));
     }
 
     private static string[] GetTableList()
```
Comment (ullet) [inline NHSChoices.StubDatabaseUtility.Tests.Integration/StubDatabaseTests.cs]: `GetTableList` method no longer appears to be used.

File: NHSChoices.StubDatabaseUtility.Tests.Integration/StubDatabaseTests.cs [hunk 2]
```diff
@@ -9,8 +9,36 @@
...
+      return sourceServerConnection;
+    }
+  }
+
   [TestFixture]
-  public class StubDatabaseTests
+  public class StubDatabaseTests : StubDatabaseTestsBase
   {
     //Not great tests - just sanity ones to make sure that nothing blows up
     // ReSharper disable AssignNullToNotNullAttribute
```
Comment (ullet) [inline NHSChoices.StubDatabaseUtility.Tests.Integration/StubDatabaseTests.cs]: Doesn't look like need this `ReSharper disable AssignNullToNotNullAttribute` in `StubDatabaseTests` any more.

File: NHSChoices.StubDatabaseUtility.Tests.Integration/StubDatabaseTests.cs [hunk 3]
```diff
@@ -37,47 +65,136 @@ public void DatabaseCopy()
...
+     
+  }
+
+  [TestFixture]
+  public class StubDatabaseCrudTests : StubDatabaseTestsBase
+  {
+    private ServerConnection _serverConnection;
+    private StubDatabaseUtility _stubDatabaseUtility;
+    //Not great tests - just sanity ones to make sure that nothing blows up
+    // ReSharper disable AssignNullToNotNullAttribute
```
Comment (ullet) [inline NHSChoices.StubDatabaseUtility.Tests.Integration/StubDatabaseTests.cs]: I'm sure this is inherited from a pre-github version, but perhaps move this closer to where required (`assembly.GetManifestResourceStream(resourceName)` in `GetTableList`), or replace with check for null, or even remove if `GetTableList` is no longer required.

File: NHSChoices.StubDatabaseUtility/StubDatabaseUtility.cs [hunk 4]
```diff
@@ -121,6 +99,43 @@ private static Database CreateBlankDb(string server, string database)
...
+        }
+      }
+
+    }
+
+    public void DeleteTableData()
+    {
+      var db = GetDatabase(_destinationServerConnection);
+
+      foreach (var insertion in _insertionList.AsEnumerable().Reverse() )
```
Comment (ullet) [inline NHSChoices.StubDatabaseUtility/StubDatabaseUtility.cs]: Was wondering why `AsEnumerable()`, then figured out `List<T>` has its own void `Reverse` method that modifies in place.  Like it; it's neater than casting to `IEnumerable<Insertion>`.

**Comment by ullet** [comment]
@neilbmclaughlin @jegglestone

Like that `StubDatabaseUtility` remembers list of insertions.  Reduces risk tearing down with wrong list of tables.

**Comment by neilbmclaughlin** [comment]
This repo is no longer under active development so 'Archiving'

**Pull request closed without merging**